src/EnergyPlus/WeatherManager.hh

```diff
@@ -1284,6 +1285,7 @@ struct WeatherManagerData : BaseGlobalStruct
           DaylightSavingIsActive(false), WFAllowsLeapYears(false), curSimDayForEndOfRunPeriod(0), Envrn(0), NumOfEnvrn(0), NumEPWTypExtSets(0),
           NumWPSkyTemperatures(0), RptIsRain(0), RptIsSnow(0), RptDayType(0), HrAngle(0.0), SolarAltitudeAngle(0.0), SolarAzimuthAngle(0.0),
           HorizIRSky(0.0), TimeStepFraction(0.0), NumSPSiteScheduleNamePtrs(0), EndDayOfMonth(12, {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}),
+          EndDayOfMonthWithLeapDay(12, {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}),
```

Comment:
Would it be smarter to make these constexpr and not bait-n-switch the number of days in Feb from "28" to "29" in EndDayOfMonthWithLeapDay? Since this new array should already have the correct data? If you are adjusting the number of days in the month based on leap year, why not just do that to the existing array EndDayOfMonth and not add the new array (which would then not be const)? Just talking out loud here.

Reply:
@rraustad no problem, I thought about a few different approaches including the ones you mentioned. I ended up going this way for a couple of reasons. First, there are a lot of completely untested functions that use EndDayOfMonth already, and they all seem to assume that even during a leap year, the return value is adjusted by 1, often by using another variable that acts as a flag called LeapYearAdd (containing either a 0 or 1). If you look at the code it is repetitive and I was going to replace it all but it would have meant adding lots of new unit tests so that I was sure I wasn't breaking anything. So that ruled out just modifying EndDayOfMonth for leap years since it might have ended up double counting leap days if I wasn't careful. Second, I assumed that if you just selected a year in RunPeriod that was a leap year it contained February 29 but it is more complicated than that and depends on the weather file containing Feb 29 and if its headers specify the year, if the user specifies in RunPeriod that "Treat Weather as Actual" is Yes, and if the "Day Of Week for Start Day" is correct for a year that is a leap year. Some special cases occur when only some of these are done. I don't know why it is so complicated, but that is how the code evolved. This just led me to the point of how do I solve this bug clearly. Honestly, all the leap year could use some rethinking and refactoring but that seemed a bigger task than just fixing the bug.

Reply:
I know why it is so complicated and I agree that it all needs to rethought and refactored, approximately one billion percent. Based on my experience working on the RunPeriod object a few years ago, the primary problem is that way the simulation tracks time (including end of day flags and the like) is shall we say "decentralized". As far as I can tell, there's no way to improve smaller portions of the code without touching all sorts of stuff that is only tangentially related. Leap days are pretty much special cases all the way down because the code had been pretty forgiving on whether or not the weather data is required (and it doesn't help that there are AMY files being sold for leap years without leap days 😱). I think I'll stop here before I start ranting about the Gregorian calendar and dominical letters.